### PR TITLE
[DOCS-120] Search indexer script

### PR DESCRIPTION
**Pull request by fonty1**
To test the indexer locally with and without DUD (lol) flag - 

`npm ci`

Goto, `src/docs/scripts/search`

`DATA_UNIONS_DOCS=on node -r esm indexer.js`

and without DUD,

`node -r esm indexer.js`

It should generate the `index.json` and `search.json` inside the platform app and become 'searchable' immediately.

Will follow this up with CI integration and tests.

### CODE REVIEW

File: app/src/docs/components/Search/searchUtils.js [hunk 1]
```diff
@@ -95,12 +95,12 @@ export const runSearchQuery = ({ query, index, store }: any) => {
             q.term(lunr.tokenizer(query), {
                 fields: ['content'],
                 boost: 100,
-                usePipeline: false,
+                usePipeline: true,
```
Comment (fonty1) [inline app/src/docs/components/Search/searchUtils.js:98]: Unrelated to the search indexer but should be set `true` to improve search results.

File: app/src/docs/scripts/search/utils.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    mdxModule.exports = fs.readFileSync(mdxFilename, 'utf8')
+}
+
+/**
+ * Sanitizer function.
+*/
+const toAnchor = (text) => (
+    text.trim()
+        .toLowerCase()
+        .replace(' & ', ' and ')
```
Comment (mondoreale) [inline app/src/docs/scripts/search/utils.js:30]:
```js
' & & & '.replace(' & ', ' and ')
// => ' and & & '
```
Maybe `.replace(/&/g, 'and')`? Spaces are dealt with in the next line.

File: app/src/docs/scripts/search/utils.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        .replace(/\s+/g, '-')
+)
+
+/**
+ * Generates formatted Module text content.
+*/
+const generateModuleRefTextContent = (name, path, content) => (
+    `${name.charAt(0).toUpperCase() + name.slice(1)} (${path.charAt(0).toUpperCase() + path.slice(1)}). ${content}`
```
Comment (mondoreale) [inline app/src/docs/scripts/search/utils.js:38]: Let's maybe define a `titleize` utility, pass a string to it and make it return one with first letter capitalized.

File: app/src/docs/scripts/search/utils.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+export async function processModuleReferenceDocs() {
+    const modules = []
+
+    canvasModules.forEach((module) => {
+        const { helpText } = module.help
+        const { path, name } = module
+
+        modules.push({
+            refContent: h2p(helpText).replace(/(\r\n|\n|\r)/gm, ''),
+            refPath: toAnchor(path.split(':')[0].trim()),
```
Comment (mondoreale) [inline app/src/docs/scripts/search/utils.js:70]: `toAnchor` trims strings already.

File: app/src/docs/scripts/search/indexer.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import {
```
Comment (mondoreale) [inline app/src/docs/scripts/search/indexer.js:1]: Any thought on moving contents of this file to `bin/indexer` & `chmod +x bin/indexer`. A couple of other changes would have to happen, i.e.:

```js
#!/usr/bin/env node -r esm

const path = require('path')

const storePath = path.resolve(__dirname, '../src/docs/components/Search/index/store.json')

const indexPath = path.resolve(__dirname, '../src/docs/components/Search/index/index.json')

// The rest of the script here. Above paths can be used to point to store and index jsons.
```

To run the indexing you'd then execute `bin/indexer` or `DATA_UNIONS_DOCS=on bin/indexer`.

File: app/src/docs/scripts/search/indexer.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            throw err
+        }
+    })
+}
+
+/**
+ * Process the docs. Save the index & store as JSON files.
+*/
+(async function start() {
+    console.warn('Generate the Docs Search Index & Store...')
```
[the remark this answers is not in the document]
Comment (mr-matto) [reply app/src/docs/scripts/search/indexer.js:36]: Good point 👍

File: app/src/docs/scripts/search/utils.js [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+const strip = require('remark-mdx-to-plain-text')
+const lunr = require('lunr')
+const h2p = require('html2plaintext')
+const canvasModules = require('../moduleReferences/canvasModuleHelpData.json')
+
+const baseModuleRefPath = '/docs/module-reference/'
+const searchStore = {}
+
+/**
+ * Injest mdx as plain text.
```
[the remark this answers is not in the document]
Comment (mr-matto) [reply app/src/docs/scripts/search/utils.js:16]: 😂 sorry! I misspelled 'ingest' which is still a terrible word choice. Will fix.

File: app/src/docs/scripts/search/indexer.js (hunk 5 above)
Comment (fonty1) [reply app/src/docs/scripts/search/indexer.js:1]: I was thinking `app/scripts` like Juha mentioned would be a good home for this, then bind it to an npm run script. Does putting it in `bin/indexer` have any benefits over that approach @mondoreale ?

**Comment by fonty1** [comment]
thanks @mondoreale  @juhah , please recast your eyes when you have time - issues addressed:
- clearer logging/docs 
- tidyup toAnchor
- added a titleize func
- moved indexer to app/scripts
- moved module references to app/scripts
- fixed bug where non-santized pathnames sneak into the store (the custom%20modules bug)

In part 3 of this trilogy I will add sentry to the script, add some more tests, add it to the npm scripts (good idea Juha) and integrate it into CI before the build step.

File: app/src/docs/scripts/search/indexer.js (hunk 5 above)
Comment (mondoreale) [reply app/src/docs/scripts/search/indexer.js]: @fonty1, benefits not really. You interact with it differently, naturally.

Script in package would look like this
```json
"indexer": "node -r esm /path/to/indexer.js",
```
Then you run it like this
```
npm run indexer
```
With `bin/indexer` you don't need a script cause you can run it directly. If you want a trace of it in package then you can also define a script
```json
"indexer": "bin/indexer",
```

At the end of the day it's what you like better.

Comment (fonty1) [reply app/src/docs/scripts/search/indexer.js]: hmm, I'll keep it where it is for now, since it's there. If there's issues I'll throw it in the `bin` :)

**Pull request merged**